### PR TITLE
Feature/vm2 annotate raises exceptions

### PR DESCRIPTION
**Pull request by mjfh**
please see issue #636

the gist of the subject is also commented at the last commit _track and annotate exceptions for VM2_

just see that the CI bails out in i386/linux, so there will be an update needed

### CODE REVIEW

**Comment by jangko** [comment]
from windows CI:

```
D:\a\nimbus-eth1\nimbus-eth1\premix\hunter.nim(73, 29) Error: can raise an unlisted exception: Exception
```

this inconsistency probably came from nim stdlib, it's a known old story. some of them already fixed in Nim 1.4.x, but we are still using Nim 1.2.x

File: nimbus/config.nim [hunk 1]
```diff
@@ -382,6 +382,9 @@ proc processCustomGenesisConfig(customGenesis: JsonNode): ConfigStatus =
     extraData = hexToSeqByte("0x")
     gasLimit = 16777216.int64
     difficulty = 1048576.u256
+
+    # note that the following parseJson() assignment implies a possible
+    # base class `Exception`
```
Comment (zah) [inline nimbus/config.nim:387]: When you work with a third-party function that has inappropriate raises list, you can usually fix the issue by wrapping it. If the raises list was incorrectly computed and the function is not expected to raise under normal circumstances, you can turn the general `Exception` type into a `Defect` through an assertion:

```nim
let
  alloc = try: parseJson("{}")
          except Exception: raiseAssert "Parsing an empty record should work"
```

If the function raises recoverable errors that are not derived from `CatchableError`, this should be filed as a bug upstream, but you can still remap the exception to a proper `CatchableError` type in a wrapper.

File: nimbus/genesis.nim [hunk 2]
```diff
@@ -118,8 +121,11 @@ proc defaultGenesisBlockForNetwork*(id: PublicNetwork): Genesis =
   result.config = publicChainConfig(id)
 
 proc toBlock*(g: Genesis, db: BaseChainDB = nil): BlockHeader =
+
+  # note: newMemoryDB() assignment may throw an Exception exception
```
Comment (zah) [inline nimbus/genesis.nim:125]: When you start adding raises annotations, it's worth getting to the bottom of the problems.

File: nimbus/vm_compile_flags.nim [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  vm0_enabled* = not evmc0_enabled and not vm2_activated
+
+  vm0_activated* =              ## set if either naitve VM variant is activated
+    vm0_enabled or evmc0_enabled
+
+
+  evmc_enabled* =               ## set if either evmc is activated
+    defined(evmc_enabled) or defined(evmc2_enabled)
+
+  relay_exception_base_class* = ## map selected Exception tyype exceptions to
```
Comment (zah) [inline nimbus/vm_compile_flags.nim:45]: I'm not sure I understand the rationale for having this mode. We should strive to have the correct raises annotations enabled by default and we don't need a mode that disables them.

File: nimbus/vm2/stack.nim [hunk 4]
```diff
@@ -9,6 +9,8 @@ import
   chronicles, strformat, strutils, sequtils, macros, eth/common, nimcrypto,
   ../errors, ../validation
 
+{.push raises: [Defect,FullStack,InsufficientStack,ValueError,ValidationError].}
```
Comment (zah) [inline nimbus/vm2/stack.nim:12]: It's better to have a very narrow raises list specified at the top-level - something that is really true for every function in the file (e.g. `raises: [Defect]` or `raises: [VMError, Defect]`). The functions that raise different exceptions can have explicit raises list specified on them.

**Comment by kdeme** [comment]
I found some of those Exception causes in hexary while working on https://github.com/status-im/nim-eth/pull/348
I'll push some in a PR soon

**Comment by kdeme** [comment]
@mjfh  https://github.com/status-im/nimbus-eth1/pull/644 should hopefully remove the raw `Exception`s from trie/hexary & trie/db

**Comment by mjfh** [comment]
outdated, zombie

**Pull request closed without merging**